booster/library/Booster/Builtin.hs

```diff
@@ -32,37 +43,240 @@ type BuiltinFunction = [Term] -> Except Text (Maybe Term)
+    | [FalseBool, _] <- args = pure $ Just TrueBool
+    | [TrueBool, FalseBool] <- args = pure $ Just FalseBool
+    | [TrueBool, TrueBool] <- args = pure $ Just TrueBool
+    | otherwise = pure Nothing -- arguments not determined
+
+boolOperator :: (Bool -> Bool -> Bool) -> BuiltinFunction
+boolOperator f args
+    | length args /= 2 =
+        throwE . renderText $ "BOOL.<operator>: wrong arity " <> pretty (length args)
+    | [Just arg1, Just arg2] <- map readBoolTerm args =
+        pure . Just . boolTerm $ f arg1 arg2
+    | otherwise = pure Nothing -- arguments not determined
+
+boolTerm :: Bool -> Term
+boolTerm True = TrueBool
+boolTerm False = FalseBool
+
+readBoolTerm :: Term -> Maybe Bool
+readBoolTerm TrueBool = Just True
+readBoolTerm FalseBool = Just False
+readBoolTerm _other = Nothing
+
+------------------------------------------------------------
+-- INT hooks
+
+builtinsINT :: Map ByteString BuiltinFunction
+builtinsINT =
+    Map.mapKeys ("INT." <>) $
+        Map.fromList
+            [ "gt" ~~> compareInt (>)
+            , "ge" ~~> compareInt (>=)
+            , "eq" ~~> compareInt (==)
+            , "le" ~~> compareInt (<=)
+            , "lt" ~~> compareInt (<)
+            , "ne" ~~> compareInt (/=)
+            , -- arithmetic
+              "add" ~~> intOperator (+)
+            , "sub" ~~> intOperator (-)
+            , "mul" ~~> intOperator (*)
+            , "abs" ~~> intModifier abs
+            -- tdiv, tmod (truncating toward zero), ediv, emod (euclidian)
+            -- bitwise operations
+            -- and, or, xor, not, shl, shr
+            -- exponentiation
+            -- pow, powmod, log2 (truncating)
+            ]
+
+compareInt :: (Integer -> Integer -> Bool) -> BuiltinFunction
+compareInt f args
+    | length args /= 2 =
+        throwE . renderText $ "INT.<comparison>: wrong arity " <> pretty (length args)
+    | [arg1, arg2] <- args
+    , Just i1 <- readIntTerm arg1
+    , Just i2 <- readIntTerm arg2 =
+        pure . Just . boolTerm $ f i1 i2
+    | otherwise = pure Nothing
+
+intOperator :: (Integer -> Integer -> Integer) -> BuiltinFunction
+intOperator f args
+    | length args /= 2 =
+        throwE . renderText $ "INT.<operator>: wrong arity " <> pretty (length args)
+    | [arg1, arg2] <- args
+    , Just i1 <- readIntTerm arg1
+    , Just i2 <- readIntTerm arg2 =
+        pure . Just . intTerm $ f i1 i2
+    | otherwise = pure Nothing
+
+intModifier :: (Integer -> Integer) -> BuiltinFunction
+intModifier f args
+    | length args /= 1 =
+        throwE . renderText $ "INT.<operator>: wrong arity " <> pretty (length args)
+    | [arg] <- args
+    , Just i <- readIntTerm arg =
+        pure . Just . intTerm $ f i
+    | otherwise = pure Nothing
+
+intTerm :: Integer -> Term
+intTerm = DomainValue SortInt . pack . show
+
+readIntTerm :: Term -> Maybe Integer
+readIntTerm (DomainValue SortInt val) = readMaybe (unpack val)
+readIntTerm _other = Nothing
```

Comment:
Obviously these hooks would be faster if we had an internal type holding an `Integer`, instead of reading and showing the values for every operation.
Currently the IMP-sum-to-n  test is still 30% faster with the LLVM backend than it is without.